lib/galaxy/webapps/galaxy/services/history_contents.py

```diff
@@ -1600,9 +1610,30 @@ def _purge(self, item: HistoryItemModel, trans: ProvidesHistoryContext):
             return self.dataset_collection_manager.delete(trans, "history", item.id, recursive=True, purge=True)
         self.hda_manager.purge(item, flush=self.flush)
 
-    def _change_datatype(self, item: HistoryItemModel, params: ChangeDatatypeOperationParams):
+    def _change_datatype(
+        self, item: HistoryItemModel, params: ChangeDatatypeOperationParams, trans: ProvidesHistoryContext
+    ):
         if isinstance(item, HistoryDatasetAssociation):
-            item.change_datatype(params.datatype)
+            self._ensure_can_change_hda_datatype(item)
+            self._set_hda_datatype(item, params.datatype, trans)
+            set_metadata.delay(dataset_id=item.id)
+
+    def _set_hda_datatype(self, hda: HistoryDatasetAssociation, datatype: str, trans: ProvidesHistoryContext):
+        if datatype == "auto":
+            path = hda.dataset.file_name
+            datatype = sniff.guess_ext(path, trans.app.datatypes_registry.sniff_order)
+        else:
+            datatype = datatype
+        trans.app.datatypes_registry.change_datatype(hda, datatype)
+        trans.sa_session.flush()
```

Comment:
Can you create a change_datatype task and push that there ? The sniffing can cause significant I/O.

Reply:
Sure! I can try

Reply:
Done! I learned a bit of celery today, thank you! :smile: